lib/Conversion/AffineToHandshake/AffineToHandshake.cpp

```diff
@@ -0,0 +1,835 @@
+    Operation *memoryOp = rewriter.create<MemoryOp>(
+        entryBlock->front().getLoc(), operands, memOpInfo.numLoadOps,
+        numAccessOps, isLoadStoreQueue, numMemoryOps, memref);
+
+    // Connect the result of this newly created MemoryOp instance, which is a
+    // memory in handshake, to the access operations.
+    setMemoryForLoadOps(accessOps, memoryOp);
+
+    // TODO: enable the control network later.
+    setMemoryControlNetwork(memoryOp, accessOps, controlValues, indexMap,
+                            memOpInfo, rewriter);
+  }
+
+  // Add ForkOp to memory results that have been used more than once.
+  forkMemoryOpResults(f, rewriter);
+
+  removeAllocOps(f, rewriter);
+  // This is called because some previously dangling results are now connected
+  // to the memory operation, and therefore, the SinkOps connected to them
+  // should be removed.
+  removeRedundantSinkOps(f, rewriter);
+}
+
+/// Remove the redundant StartOp from the entry block of f. We need to remove
+/// this StartOp because the control value that it produces becomes the last
+/// block argument of the entry block. Therefore, the StartOp becomes redundant.
+void rewriteStartOp(handshake::FuncOp f, ConversionPatternRewriter &rewriter) {
+  Block *entryBlock = &f.front();
+
+  // Get the last block argument and replace every use of the StartOp's result
+  // with it.
+  mlir::Value val = entryBlock->getArguments().back();
+  Operation *startOp = getStartOp(entryBlock);
+  assert(startOp != nullptr);
+  startOp->getResult(0).replaceAllUsesWith(val);
+
+  // Remove the StartOp.
+  rewriter.eraseOp(startOp);
+}
+
+/// Lowering from standard FuncOp to handshake FuncOp. The rewrite function
+/// does the following things: Creating a new handshake::FuncOp from the
+/// standard mlir::FuncOp, rewriting a bunch of operations within the
+/// mlir::FuncOp, and does some analysis.
+class FuncOpLowering : public OpConversionPattern<mlir::FuncOp> {
+public:
+  using OpConversionPattern<mlir::FuncOp>::OpConversionPattern;
+
+  /// Get the attributes for the handshake::FuncOp from the original
+  /// mlir::FuncOp. Only symbol and type attributes are kept.
+  void getFuncOpAttributes(mlir::FuncOp funcOp,
+                           SmallVectorImpl<NamedAttribute> &attrs) const {
+    for (const auto &attr : funcOp.getAttrs()) {
+      if (attr.first == SymbolTable::getSymbolAttrName() ||
+          attr.first == impl::getTypeAttrName())
+        continue;
+      attrs.push_back(attr);
+    }
+  }
+
+  /// Get the types of the arguments and results of mlir::FuncOp. A None type
+  /// will be appended to both arguments and results lists to represent
+  /// control input/output. The rewriter argument will be used to generate
+  /// that None type object.
+  void getArgAndResultTypes(mlir::FuncOp funcOp,
+                            SmallVectorImpl<mlir::Type> &argTypes,
+                            SmallVectorImpl<mlir::Type> &resTypes,
```

Comment:
ArrayRef?

Reply:
Thanks Steve. I've tried that, but it seems that `ArrayRef` is a constant reference to the array, while I have to update the content of the array. Also, I need access to `std::vector` APIs like `resize`.